### PR TITLE
[LLD][COFF] Support /DEPENDENTLOADFLAG[:flags]

### LINKED ISSUE #43935: Add /DEPENDENTLOADFLAG linker flag to lld-link
|  |  |
| --- | --- |
| Bugzilla Link | [44590](https://llvm.org/bz44590) |
| Version | unspecified |
| OS | Windows NT |
| CC | @froydnj,@rnk,@smithp35 |

## Extended Description 
This is useful to mitigate DLL preload attacks, but currently only MS link.exe supports it.

### PR DESCRIPTION
**Pull request by nurmukhametov**
This should fix https://github.com/llvm/llvm-project/issues/43935

### CODE REVIEW

File: lld/test/COFF/dependentflags.test [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+RUN: not lld-link %S/Inputs/precomp-a.obj %t.ldcfg.obj /out:%t.exe /nodefaultlib /force /dependentloadflag:zz 2>&1 | FileCheck %s --check-prefix FAIL
+RUN: not lld-link %S/Inputs/precomp-a.obj %t.ldcfg.obj /out:%t.exe /nodefaultlib /force /dependentloadflag:0xf0000 2>&1 | FileCheck %s --check-prefix FAIL-RANGE
+
+
+BASE: DependentLoadFlags: 0x0
+FLAGS: DependentLoadFlags: 0x800
+
+FAIL: lld-link: error: /dependentloadflag: invalid argument: zz
+FAIL-RANGE: lld-link: error: /dependentloadflag: invalid argument: 0xf0000
```
Comment (tru) [inline lld/test/COFF/dependentflags.test:28]: Missing EOF.

File: lld/test/COFF/dependentflags.test [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+// ---- Many arguments
+RUN: lld-link %S/Inputs/precomp-a.obj %t.ldcfg.obj /out:%t.exe /nodefaultlib /force /dependentloadflag:0x800 /dependentloadflag
+RUN: llvm-readobj --coff-load-config %t.exe | FileCheck %s --check-prefix BASE
+
+RUN: lld-link %S/Inputs/precomp-a.obj %t.ldcfg.obj /out:%t.exe /nodefaultlib /force /dependentloadflag /dependentloadflag:0x800
+RUN: llvm-readobj --coff-load-config %t.exe | FileCheck %s --check-prefix FLAGS
+
+RUN: not lld-link %S/Inputs/precomp-a.obj %t.ldcfg.obj /out:%t.exe /nodefaultlib /force /dependentloadflag:zz 2>&1 | FileCheck %s --check-prefix FAIL
+RUN: not lld-link %S/Inputs/precomp-a.obj %t.ldcfg.obj /out:%t.exe /nodefaultlib /force /dependentloadflag:0xf0000 2>&1 | FileCheck %s --check-prefix FAIL-RANGE
```
Comment (tru) [inline lld/test/COFF/dependentflags.test:21]: might want to have a test to make sure the last argument is the one being used.

File: lld/COFF/Writer.cpp [hunk 3]
```diff
@@ -696,6 +699,7 @@ void Writer::run() {
       writeHeader<pe32_header>();
     }
     writeSections();
+    changeLoadConfig();
```
Comment (tru) [inline lld/COFF/Writer.cpp:702]: Should this run even if `dependentloadflag` is not set?

**Review by tru — COMMENTED** [review]
I am not good with the internals of what should change here, but here are some smaller nit comments.

**Comment by tru** [comment]
Also make sure to run clang-format to fix the formatting problems.

File: lld/COFF/Writer.cpp [hunk 4]
```diff
@@ -2256,6 +2262,32 @@ void Writer::fixTlsAlignment() {
...
+    changeLoadConfigGuardData(
+        reinterpret_cast<coff_load_configuration64 *>(symBuf));
+  else
+    changeLoadConfigGuardData(
+        reinterpret_cast<coff_load_configuration32 *>(symBuf));
+}
+
+template <typename T> void Writer::changeLoadConfigGuardData(T *loadConfig) {
+  if (ctx.config.dependentLoadFlags)
+    loadConfig->DependentLoadFlags = ctx.config.dependentLoadFlags;
```
Comment (aganea) [inline lld/COFF/Writer.cpp:2288]: Quite interestingly `dependentLoadFlags` is a WORD in the loadconfig, but it’s a DWORD in `LoadLibraryEx`. Nothing to do here, this is just a remark.

File: lld/COFF/Writer.cpp [hunk 5]
```diff
@@ -2256,6 +2262,32 @@ void Writer::fixTlsAlignment() {
   }
 }
 
+void Writer::changeLoadConfig() {
```
[the remark this answers is not in the document]
Comment (nurmukhametov) [reply lld/COFF/Writer.cpp]: Yes. I have changed these duplication to `prepareLoadConfig`. I hope it is what you meant.

File: lld/COFF/Writer.cpp (hunk 4 above)
Comment (nurmukhametov) [reply lld/COFF/Writer.cpp]: Yes, although it looks like that the higher part of DWORD is not used at the moment ([link](https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexw#parameters)).

File: lld/COFF/DriverUtils.cpp [hunk 6]
```diff
@@ -265,6 +265,19 @@ void LinkerDriver::parseFunctionPadMin(llvm::opt::Arg *a) {
...
+void LinkerDriver::parseDependentLoadFlags(llvm::opt::Arg *a) {
+  StringRef arg = a->getNumValues() ? a->getValue() : "";
+  if (!arg.empty()) {
+    if (arg.getAsInteger(0, ctx.config.dependentLoadFlags))
+      error("/dependentloadflag: invalid argument: " + arg);
+    return;
+  }
+  // MSVC linker reports error "no argument specified", although MSDN describes
+  // argument as optional.
+  error("no argument specified with option '/dependentloadflag'");
```
Comment (aganea) [inline lld/COFF/DriverUtils.cpp:278]: I think this error can be made a bit more concise. Also other errors for other flags seem to put the flag in the front: "/dependentloadflag: no argument specified"

**Review by aganea — APPROVED** [review]
Looks good to me otherwise! Thank you!

File: lld/COFF/DriverUtils.cpp (hunk 6 above)
Comment (nurmukhametov) [reply lld/COFF/DriverUtils.cpp]: Done

File: lld/COFF/Writer.cpp (hunk 3 above)
Comment (nurmukhametov) [reply lld/COFF/Writer.cpp]: Not really, it is addressed in the latest revision.

**Review by tru — APPROVED** [review]
LGTM

**Comment by nurmukhametov** [comment]
I've fixed type in commit message.

**Comment by luporl** [comment]
It looks like this broke clang-arm64-windows-msvc buildbot:
https://lab.llvm.org/buildbot/#/builders/65/builds/11819

**Comment by nurmukhametov** [comment]
> It looks like this broke clang-arm64-windows-msvc buildbot: https://lab.llvm.org/buildbot/#/builders/65/builds/11819

Indeed, I probably had to provide `# REQUIRES: x86` in lit tests?

**Comment by aganea** [comment]
@nurmukhametov Do you have commit rights? If yes you can push a fix commit directly without a PR.

**Comment by nurmukhametov** [comment]
> @nurmukhametov Do you have commit rights? If yes you can push a fix commit directly without a PR.

No, I don't think I have.

**Comment by aganea** [comment]
Fixed in b371ea35c16e5ca42237012ef1a935a90bcce5e9

**Comment by MaskRay** [comment]
> This should fix https://github.com/llvm/llvm-project/issues/43935

The description is too brief. I'd say "Close #43935: support /DEPENDENTLOADFLAG , which does XXX".

**Pull request merged**